validations/src/validations.rs

```diff
@@ -813,32 +820,45 @@ pub fn validate_tally_outputs<S: ::std::hash::BuildHasher>(
     dr_state: &DataRequestState,
     ta_tx: &TallyTransaction,
     n_reveals: usize,
+    n_commits: usize,
     dishonest_pkhs: HashSet<PublicKeyHash, S>,
 ) -> Result<(), failure::Error> {
     let witnesses = dr_state.data_request.witnesses as usize;
     let dishonest_len = dishonest_pkhs.len();
-    let change_required = dishonest_len > 0;
+    let change_required = dishonest_len > 0 || n_commits == 0;
 
-    if change_required && (ta_tx.outputs.len() != witnesses - dishonest_len + 1) {
+    if n_commits == 0 && (ta_tx.outputs.len() != 1) {
+        return Err(TransactionError::WrongNumberOutputs {
+            outputs: ta_tx.outputs.len(),
+            expected_outputs: 1,
+        }
+        .into());
+    } else if dishonest_len > 0 && (ta_tx.outputs.len() != witnesses - dishonest_len + 1) {
         return Err(TransactionError::WrongNumberOutputs {
             outputs: ta_tx.outputs.len(),
             expected_outputs: witnesses - dishonest_len + 1,
         }
         .into());
-    } else if !change_required && (ta_tx.outputs.len() != witnesses - dishonest_len) {
+    } else if !change_required && (ta_tx.outputs.len() != witnesses) {
         return Err(TransactionError::WrongNumberOutputs {
             outputs: ta_tx.outputs.len(),
-            expected_outputs: witnesses - dishonest_len,
+            expected_outputs: witnesses,
         }
         .into());
     }
 
     let mut pkh_rewarded: HashSet<PublicKeyHash> = HashSet::default();
-    let reveal_reward = dr_state.data_request.witness_reward;
+    let witness_reward = dr_state.data_request.witness_reward;
+    let reveal_fee = dr_state.data_request.reveal_fee;
+    let commit_fee = dr_state.data_request.commit_fee;
     for (i, output) in ta_tx.outputs.iter().enumerate() {
         if change_required && i == ta_tx.outputs.len() - 1 && output.pkh == dr_state.pkh {
-            let expected_tally_change = reveal_reward * dishonest_len as u64
-                + dr_state.data_request.reveal_fee * (witnesses - n_reveals) as u64;
+            let expected_tally_change = if n_reveals == 0 {
+                witnesses as u64 * (witness_reward + reveal_fee + commit_fee)
```

Comment:
Is this correct when `n_reveals == 0` but `n_commits > 0`?

Reply:
O_O my mistake, I modified in other part but forgot here!! Thanks!!